[Skia][Cairo] convertImagePixels copies wrongly if no pixel conversion is needed and bytesPerRow are different for source and destination

fujii opened the pull request:
#### 1de35d8ca5de3b5707f2b0b23367af05e32327fe
<pre>
[Skia][Cairo] convertImagePixels copies wrongly if no pixel conversion is needed and bytesPerRow are different for source and destination
<a href="https://bugs.webkit.org/show_bug.cgi?id=286664">https://bugs.webkit.org/show_bug.cgi?id=286664</a>

Reviewed by Sam Weinig.

&lt;<a href="https://commits.webkit.org/241230@main">https://commits.webkit.org/241230@main</a>&gt; added a fast code path of
convertImagePixels for Cairo port in case of no pixel conversion was
needed. However, it didn&apos;t take bytesPerRow and destinationSize into
account. It didn&apos;t work as expected.

&lt;<a href="https://commits.webkit.org/277802@main">https://commits.webkit.org/277802@main</a>&gt; added a same code for Skia.

* Source/WebCore/platform/graphics/PixelBufferConversion.cpp:
(WebCore::copyImagePixels):
(WebCore::convertImagePixels):
* Source/WebCore/platform/graphics/PixelBufferConversion.h:
* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/Tests/WebCore/PixelBufferConversionTests.cpp: Added.
(TestWebKitAPI::TEST(PixelBufferConversionTests, convertImagePixels)):
(TestWebKitAPI::TEST(PixelBufferConversionTests, convertImagePixels2)):

Canonical link: <a href="https://commits.webkit.org/289520@main">https://commits.webkit.org/289520@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2fc2cdb04e261df9d0a4e94b158250251af100b4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87163 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6673 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41523 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92024 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37902 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89213 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6949 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14742 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67365 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25102 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90165 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5319 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78892 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47685 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5092 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33270 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37019 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75576 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34147 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93909 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14325 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10417 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76166 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14529 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74748 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75368 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19710 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18149 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7242 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13585 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14344 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/19637 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14089 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17532 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15870 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Tools/TestWebKitAPI/Tests/WebCore/PixelBufferConversionTests.cpp:
```diff
@@ -0,0 +1,84 @@
+
+    EXPECT_EQ(convert(PixelFormat::RGBA8, PixelFormat::RGBA8), (std::vector<uint8_t> { 1, 2, 3, 4 }));
+    EXPECT_EQ(convert(PixelFormat::RGBA8, PixelFormat::BGRA8), (std::vector<uint8_t> { 3, 2, 1, 4 }));
+#if !USE(SKIA) // convertImagePixelsSkia doesn't support this
```
We should add a SUPPORTS(PIXEL_FORMAT_BGRX8) macro for this.

fujii replied:
It's easy to add BGRX8 support for Skia port just by falling back to the common CPU code.
https://github.com/WebKit/WebKit/blob/3f975ad6e201b37a19f7ed0af4473872caf4898d/Source/WebCore/platform/graphics/PixelBufferConversion.cpp#L311-L349
I will do it as a follow-up.

weinig reviewed (approved):
Thanks for adding the test!

fujii committed:
[Skia][Cairo] convertImagePixels copies wrongly if no pixel conversion is needed and bytesPerRow are different for source and destination
https://bugs.webkit.org/show_bug.cgi?id=286664

Reviewed by Sam Weinig.

<https://commits.webkit.org/241230@main> added a fast code path of
convertImagePixels for Cairo port in case of no pixel conversion was
needed. However, it didn't take bytesPerRow and destinationSize into
account. It didn't work as expected.

<https://commits.webkit.org/277802@main> added a same code for Skia.

* Source/WebCore/platform/graphics/PixelBufferConversion.cpp:
(WebCore::copyImagePixels):
(WebCore::convertImagePixels):
* Source/WebCore/platform/graphics/PixelBufferConversion.h:
* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/Tests/WebCore/PixelBufferConversionTests.cpp: Added.
(TestWebKitAPI::TEST(PixelBufferConversionTests, convertImagePixels)):
(TestWebKitAPI::TEST(PixelBufferConversionTests, convertImagePixels2)):

Canonical link: https://commits.webkit.org/289520@main
